firedrake/ensemble.py

```diff
@@ -0,0 +1,65 @@
+from pyop2.mpi import MPI
+
+__all__ = ("Ensemble", )
+
+
+class Ensemble(object):
+    def __init__(self, comm, M):
+        """
+        Create a set of space and ensemble subcommunicators.
+
+        :arg comm: The communicator to split.
+        :arg M: the size of the communicators used for spatial parallelism.
+        :raises ValueError: if ``M`` does not divide ``comm.size`` exactly.
+        """
+        size = comm.size
+
+        if (size // M)*M != size:
+            raise ValueError("Invalid size of subcommunicators %d does not divide %d" % (M, size))
+
+        rank = comm.rank
+
+        self.gcomm = comm
+        """The global communicator."""
+
+        self.comm = comm.Split(color=(rank // M), key=rank)
+        """The communicator for spatial parallelism, contains a
+        contiguous chunk of M processes from :attr:`comm`"""
+
+        self.ecomm = comm.Split(color=(rank % M), key=rank)
+        """The communicator for ensemble parallelism, contains all
+        processes in :attr:`comm` which have the same rank in
+        :attr:`scomm`."""
+
+        self.erank = self.ecomm.rank
```

Comment:
I'm not convinced about the abbreviations here. I think as a matter of policy that readability is more important than brevity. I'd vote for `global_comm`, `ensemble_comm`. I'm also not convinced we need `self.erank`. Surely self.ensemble_comm.rank is clearer.